nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,378 @@
+		}
+	}
+	err := uploadUpgradeBundle(m, upgradeBundleURL, userName, password, upgradeBundleType, version)
+	if err != nil {
+		return fmt.Errorf("Failed to upload upgrade bundle: %s", err)
+	}
+	return nil
+}
+
+func precheckBundleCompatibilityCheck(precheckBundleURL string) bool {
+	if nsxVersionLower("4.1.1") && len(precheckBundleURL) > 0 {
+		return false
+	}
+	if nsxVersionHigherOrEqual("4.1.1") && len(precheckBundleURL) == 0 {
+		return false
+	}
+	return true
+}
+
+func uploadUpgradeBundle(m interface{}, url string, userName string, password string, bundleType string, version string) error {
+	connector := getPolicyConnector(m)
+	client := upgrade.NewBundlesClient(connector)
+	bundleFetchRequest := nsxModel.UpgradeBundleFetchRequest{
+		Url: &url,
+	}
+	if nsxVersionHigherOrEqual("4.1.1") {
+		bundleFetchRequest.BundleType = &bundleType
+		bundleFetchRequest.Password = &password
+		bundleFetchRequest.Username = &userName
+		bundleFetchRequest.Version = &version
+	}
+	bundleID, err := client.Create(bundleFetchRequest, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to upload upgrade bundle of type %s: %v", bundleType, err)
+	}
+	return waitForBundleUpload(m, *bundleID.BundleId)
+}
+
+func acceptUserAgreement(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	acceptUserAgreement := d.Get("accept_user_agreement").(bool)
+	if !acceptUserAgreement {
+		return fmt.Errorf("To proceed with upgrade, you must accept user agreement")
+	}
+	client := eula.NewAcceptClient(connector)
+	err := client.Create()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func upgradeUc(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewUpgradeClient(connector)
+	err := client.Upgradeuc()
+	if err != nil {
+		return err
+	}
+	return waitForUcUpgrade(m)
+}
+
+func executePreupgradeChecks(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewUpgradeClient(connector)
+	err := client.Executepreupgradechecks(nil, nil, nil, nil, nil, nil)
+	if err != nil {
+		return err
+	}
+	for _, componentType := range precheckComponentTypes {
+		err = waitForPrecheckComplete(m, componentType)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getPrecheckErrors(m interface{}, typeParam string) ([]nsxModel.UpgradeCheckFailure, error) {
+	connector := getPolicyConnector(m)
+	client := pre_upgrade_checks.NewFailuresClient(connector)
+	resultList, err := client.List(nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, &typeParam, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	return resultList.Results, nil
+}
+
+func setFailedPrechecksInSchema(d *schema.ResourceData, precheckWarnings []nsxModel.UpgradeCheckFailure) error {
+	var failedPrechecksList []map[string]interface{}
+	for _, warning := range precheckWarnings {
+		if *warning.NeedsAck {
+			id := *warning.Id
```

Comment:
I think we should list all warnings, including those that don't require ack. Maybe we can list them in separate arrays for clarity.

Reply:
done